### DISCUSSION TITLE
Tracker: Tools

### DISCUSSION THREAD

**Discussion by novhack** [Orama-Interactive/Pixelorama#592] [Ideas]
The following list also contains some suggestions from me that come from what is available in other pixel art tools.
I also went through issues and compiled the ones that are related to tools.

### Pencil
- [ ] #574
- [ ] #569

### Shape tools
- [x] #359 
- [x] `Shift` to make a circle or square
- [x] `Ctrl` to move center of shape to click origin

### Line tool
- [x] A dedicated line tool
Could have the v0.7 way of drawing to make it different from pencil `shift` mode

### Curve tool
- [x] A dedicated curve tool. EDIT: Implemented in #1019.
- Graphics Gale has a very nice implementation of this
- Aseprite and MSPaint have a little simpler implementation
![2020-10-23-20-39-12](https://user-images.githubusercontent.com/42614907/97041759-43073b80-1570-11eb-8220-45b6186381c2.gif)
![2020-10-23-20-40-18](https://user-images.githubusercontent.com/42614907/97041763-46022c00-1570-11eb-9568-b265d5ce3800.gif)

### Polygon tool
- [x] Advanced tool in Aseprite that seems useful. EDIT: The curve tool has a fill shape option, which can be used to make filled polygons. Is a dedicated polygon tool needed?
![2020-10-23-20-19-39](https://user-images.githubusercontent.com/42614907/97040338-12be9d80-156e-11eb-9cc5-0238e53bf27f.gif)

### Contour tool
- [x] Implemented in #489 which is based on #459
This one is in Aseprite
![2020-10-23-20-28-16](https://user-images.githubusercontent.com/42614907/97040570-69c47280-156e-11eb-90dc-d667fb087125.gif)

### Text tool 
- [x] #589 (Implemented in #1134, destructive-only)

### Move tool 
- [x] #313 
- [ ] `Click + drag` could move the whole layer
- [ ]  Option for looping so pixels moved over the edge appear on other side of canvas
- Useful in animation

### Zoom tool
- [x] #211
- [x] It could be more useful to react to mouse `click + drag`
    - Drag left from click origin -> zoom out
    - Drag right from click origin -> zoom in

### Advanced selection tool
- [x] Rectangular, Elliptical selection
- [x] Magic wand #567
- [x] Lasso #567
- Other related issues:
	- [ ] #56
	- [ ] #563 (rotate selection?)

### Lighten/darken
- [x] #189

### Other tools related issues:
- [ ] #577
- [ ] #575
- [x] #125

**Comment by OverloadedOrama** [comment]
The contour tool seems pretty straightforward. If I'm not wrong, it's basically the pencil tool, but after you release the mouse it connects the end point to the start point and then just fills the area. I think it might be better if it was a pencil option, like "Contour mode" instead of a tool of its own.

**Comment by novhack** [comment]
> then just fills the area.

This part might be a little bit more complicated with overlapping lines but the idea of adding it into Pencil seems alright.

**Comment by OverloadedOrama** [comment]
Leaving this for future reference for ellipses and Bézier curves: http://members.chello.at/easyfilter/bresenham.html

**Comment by Variable-ind** [comment]
The text tool is added in 1.1 (list should now be updated 😁)
